### PR TITLE
executor: fix bug that pessimistic lock doesn't work on a partition table when index merge is used

### LINKED ISSUE #31024: Pessimistic lock doesn't work on a partition table when index merge is used
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table if exists t1, t2;
    create table t1 (c_datetime datetime, c1 int, c2 int, primary key (c_datetime), key(c1), key(c2)) partition by range (to_days(c_datetime)) ( partition p0 values less than (to_days('2020-02-01')), partition p1 values less than (to_days('2020-04-01')), partition p2 values less than           (to_days('2020-06-01')), partition p3 values less than maxvalue) ;
    create table t2 (c_datetime datetime, unique key(c_datetime));
    insert into t1 values ('2020-06-26 03:24:00', 1, 1), ('2020-02-21 07:15:33', 2, 2), ('2020-04-27 13:50:58', 3, 3);
    insert into t2 values ('2020-01-10 09:36:00'), ('2020-02-04 06:00:00'), ('2020-06-12 03:45:18');
    SET GLOBAL tidb_txn_mode = 'pessimistic';

txn1:

    begin;
    select /*+ use_index_merge(t1) */ * from t1 join t2 on t1.c_datetime >= t2.c_datetime where t1.c1 < 10 or t1.c2 < 10 for update; 

txn2:

    begin;
    update t1 set c_datetime = '2020-06-26 03:24:00' where c1 = 1;

### 2. What did you expect to see? (Required)
txn2 will be blocked.
### 3. What did you see instead (Required)
txn2 didn't be blocked.

    mysql> begin;
    Query OK, 0 rows affected (0.01 sec)
    mysql> update t1 set c_datetime = '2020-06-26 03:24:00' where c1 = 1;
    Query OK, 0 rows affected (0.01 sec)
    Rows matched: 1  Changed: 0  Warnings: 0

### 4. What is your TiDB version? (Required)

    mysql> select tidb_version();
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | tidb_version()
                                    |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
    | Release Version: v5.4.0-alpha-498-g404895c04a-dirty
    Edition: Community
    Git Commit Hash: 404895c04a5183d6bf759cc910e642475f4f9f2b
    Git Branch: master
    UTC Build Time: 2021-12-25 06:18:03
    GoVersion: go1.16.4
    Race Enabled: false
    TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
    Check Table Before Drop: false |
    +----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #31024

Problem Summary: just as #31024 said.

### What is changed and how it works?

Related to #14921 and #21148.

We need to add `extraPIDColumnIndex` when build final TableReader for IndexMerge, just like #21148 did.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
fix bug that pessimistic lock doesn't work on a partition table when IndexMerge is used
```

### CODE REVIEW

**Review by tiancaiamao — APPROVED** [review]
I'm planning to fix partition table lock issue using alternative way https://github.com/pingcap/tidb/pull/30732

The current `extraPIDColumn` way is buggy and I've abandoned it, see  https://github.com/pingcap/tidb/pull/28666 for another fix on UnionScan

This PR still looks good to me because it add a test and is helpful to prevent regression at least.

**Comment by guo-shaoge** [comment]
Guess it's ok to wait for the final fix instead of this temporary fix.

File: executor/index_merge_reader_test.go [hunk 1]
```diff
@@ -454,3 +455,47 @@ func (test *testSerialSuite2) TestIndexMergeReaderMemTracker(c *C) {
...
+	tk1 := testkit.NewTestKit(c, test.store)
+	tk1.MustExec("use test")
+
+	tk.MustExec("begin pessimistic")
+	tk.MustQuery(`select /*+ use_index_merge(t1) */ c1 from t1 join t2
+			on t1.c_datetime >= t2.c_datetime
+			where t1.c1 < 10 or t1.c2 < 10 for update;`).Sort().Check(testkit.Rows("1", "1", "1", "2", "2", "3", "3"))
+	tk1.MustExec("begin pessimistic")
+
+	ch := make(chan int32, 5)
```
[the remark this answers is not in the document]
Comment (guo-shaoge) [reply executor/index_merge_reader_test.go:484]: IMHO, 1 is also ok.

**Review by mjonss — APPROVED** [review]
LGTM

This will fix the issue in static prune mode.

**Pull request closed without merging**